File: 2019/day11.go
package main

import (
	"github.com/lukechampine/advent/2019/intcode"
	"github.com/lukechampine/advent/utils"
)

var input = utils.ReadInput("day11_input.txt")
var prog = utils.ExtractInts(input)

const (
	Black = 0
	White = 1
)

func main() {
	// part 1
	m := intcode.New(prog)

	robot := utils.NewAgent(0, 0, utils.Up)
	panels := make(map[utils.Pos]int)
	getInput := func() int {
		if panels[robot.Pos] == White {
			return 1
		}
		return 0
	}
	paintPanel := func(color int) {
		switch color {
		case 0:
			panels[robot.Pos] = Black
		case 1:
			panels[robot.Pos] = White
		default:
			panic("unreachable")
		}
	}
	turn := func(dir int) {
		switch dir {
		case 0:
			robot.TurnLeft()
		case 1:
			robot.TurnRight()
		default:
			panic("unreachable")
		}
		robot.MoveForward(1)
	}
	out := m.Run(getInput())
	for !m.Halted {
		paintPanel(out[0])
		turn(out[1])
		out = m.Run(getInput())
	}
	utils.Println(len(panels))

	// part 2
	m = intcode.New(prog)
	robot = utils.NewAgent(0, 0, utils.Up)
	panels = make(map[utils.Pos]int)
	panels[utils.Pos{0, 0}] = White

	out = m.Run(getInput())
	for !m.Halted {
		paintPanel(out[0])
		turn(out[1])
		out = m.Run(getInput())
	}
	var painted []utils.Pos
	for p, color := range panels {
		if color == White {
			painted = append(painted, p)
		}
	}
	utils.PrintPositions(painted, ' ', '█')
}

File: 2019/day11.zig
const std = @import("std");
const utils = @import("utils.zig");

const Color = enum {
    Black,
    White,
};

const Panels = std.AutoHashMap(utils.Pos, Color);

const robot = struct {
    const Self = @This();
    a: utils.Actor = undefined,
    m: utils.Machine = undefined,
    panels: *Panels = undefined,

    fn run(s: *Self) bool {
        var in: i64 = switch (s.panels.getValue(s.a.pos) orelse Color.Black) {
            Color.Black => i64(0),
            Color.White => 1,
        };
        var out = s.m.run(in) orelse return false;
        _ = s.panels.put(s.a.pos, @intToEnum(Color, @intCast(u1, out[0]))) catch unreachable;
        switch (out[1]) {
            0 => s.a.turnLeft(),
            1 => s.a.turnRight(),
            else => unreachable,
        }
        s.a.moveForward(1);
        return true;
    }

    fn init(m: utils.Machine, panels: *Panels) Self {
        var a = utils.Actor{
            .pos = utils.Pos{ .x = 0, .y = 0 },
            .dir = utils.Dir.Up,
        };
        return Self{ .a = a, .m = m, .panels = panels };
    }
};

pub fn main() void {
    var prog = utils.loadProgram("day11_input.txt");

    // part 1
    var m = utils.Machine.init(prog);
    var panels = utils.makeMap(utils.Pos, Color);
    var r = robot.init(m, &panels);
    while (r.run()) {}
    utils.println(panels.count());

    // part 2
    m = utils.Machine.init(prog);
    panels = utils.makeMap(utils.Pos, Color);
    _ = panels.put(utils.Pos{ .x = 0, .y = 0 }, Color.White) catch unreachable;
    r = robot.init(m, &panels);
    while (r.run()) {}
    var minPos = utils.Pos{ .x = 99999, .y = 99999 };
    var maxPos = utils.Pos{ .x = -99999, .y = -99999 };
    var pit = panels.iterator();
    while (pit.next()) |o| {
        minPos.x = std.math.min(minPos.x, o.key.x);
        minPos.y = std.math.min(minPos.y, o.key.y);
        maxPos.x = std.math.max(maxPos.x, o.key.x);
        maxPos.y = std.math.max(maxPos.y, o.key.y);
    }
    var p = utils.Pos{ .y = maxPos.y };
    while (p.y >= minPos.y) : (p.y -= 1) {
        p.x = minPos.x;
        while (p.x <= maxPos.x) : (p.x += 1) {
            var c = panels.getValue(p) orelse Color.Black;
            utils.print(switch (c) {
                Color.Black => " ",
                Color.White => "█",
            });
        }
        utils.print("\n");
    }
}
